docs/experiments/extract-mode.md

```diff
@@ -0,0 +1,103 @@
+---
+id: extract-mode
+append_help_link: true
+description: "This article explains the extract mode, which allows for easier handling of files containing more than one language."
+---
+
+# Extract mode
+
+## Introduction
+
+Extract mode enables you to run existing rules on subsections of files where the rule language is different than the language of the file. For example, running a Javascript rule on code contained inside of script tags in an HTML document.
+
+## Example of extract mode
+
+Without extract mode, writing rules to validate template, Markdown or configuration files which contain code in another language can be burdensome and require significant rule duplication.
+
+Let's take the following Bash rule as an example (a simplified version of [our `curl-eval` rule](https://github.com/returntocorp/semgrep-rules/blob/release/bash/curl/security/curl-eval.yaml)):
+
+```yaml
+rules:
+  - id: curl-eval
+    severity: WARNING
+    languages:
+      - bash
+    message: Evaluating data from a `curl` command is unsafe.
+    mode: taint
+    pattern-sources:
+      - pattern: |
+          $(curl ...)
+      - pattern: |
+          `curl ...`
+    pattern-sinks:
+      - pattern: eval ...
+```
+
+Usually, this rule would only be run against Bash files. However, our project might contain Dockerfiles or Python scripts which invoke Bash commands&mdash;without an extract mode rule Semgrep won't know about any contained Bash, and therefore, will **not** run any Bash rules against those contained commands.
+
+However, with extract mode we can provide Semgrep with instructions on how to extract any Bash commands we might be using in a Docker `RUN` instruction or as an argument to Python's `os.system` standard library function.
+
+```yaml
+rules:
+  - id: extract-docker-run-to-bash
+    mode: extract
+    languages:
+      - dockerfile
+    pattern: RUN $...CMD
+    extract: $...CMD
+    dest-language: bash
+  - id: extract-python-os-system-to-bash
+    mode: extract
+    languages:
+      - python
+    pattern: os.system("$CMD")
+    extract: $CMD
+    dest-language: bash
+```
+
+By adding the extract mode rules show above, Semgrep can match the Bash contained in the below Python file, and will report the contained Bash as  matching against the `curl-eval` rule.
+
+```python
+from os import system
+
+if system('eval `curl -s "http://www.very-secure-website.net"`'):
+    print("Command failed!")
+else:
+    print("Success")
+```
+
+Likewise, if our query included a Dockerfile with an equivalent Bash command, such as the example below, Semgrep would report the contained Bash as matching against the `curl-eval` rule.
```

Comment:
Blocking:
- Avoid using would: <img width="893" alt="Screenshot 2022-07-12 at 23 25 50" src="https://user-images.githubusercontent.com/26669307/178598266-4956d886-bb68-4a17-8949-c9b75ff80aa8.png"> (See [would](https://developers.google.com/style/word-list?hl=en#would) in Google developer style guide)
- Do not use "we", "our", and similar words unless necessary. See: [Use of we](https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#39871bd6aca04e37bac6dc2e70dd2baf) in r2c technical communication style guide.

```suggestion
Likewise, if a query included Dockerfile with an equivalent Bash command, Semgrep reports the contained Bash as matching against the `curl-eval` rule. See the following Dockerfile example that contains a Bash command:
```

Reply:
Most of this looks good, but I believe it should still be "a Dockerfile". To the best of my knowledge, Docker doesn't call the "language" Dockerfiles are written in anything specific, and just refers to it as a text format, so I believe this ought to refer to a specific file. See, e.g., the usage https://docs.docker.com/engine/reference/builder/.

Reply:
[migrated to other branch]